src/Routing/PushState.purs

```diff
@@ -0,0 +1,247 @@
+-- | used as an implicit prefix for the paths supplied to the instance, being
+-- | automatically added and stripped as necessary.
+makeInterface :: forall eff. Basename -> Eff (PushStateEffects eff) (PushStateInterface (PushStateEffects eff))
+makeInterface basename = do
+  argsRef <- newRef []
+  freshRef <- newRef 0
+  listenersRef <- newRef StrMap.empty
+
+  let
+    notify ev =
+      traverse_ (_ $ ev) =<< readRef listenersRef
+
+    listen k = do
+      fresh <- readRef freshRef
+      writeRef freshRef (fresh + 1)
+      modifyRef listenersRef $ StrMap.insert (show fresh) k
+      pure $ modifyRef listenersRef $ StrMap.delete (show fresh)
+
+    locationState = do
+      loc <- DOM.window >>= DOM.location
+      state <- DOM.window >>= DOM.history >>= DOM.state
+      pathname <- fromMaybe <*> String.stripPrefix (String.Pattern (unwrap basename)) <$> DOM.pathname loc
+      search <- DOM.search loc
+      hash <- DOM.hash loc
+      let path = pathname <> search <> hash
+      pure { state, pathname, search, hash, path }
+
+  -- The hashchange interface is asynchronous, since hashchange events are
+  -- fired on the next tick of the event loop. We want the push-state
+  -- interface to behave as similarly as possible, so we use the microtask
+  -- queue via MutationObserver to schedule callbacks. Alternatively we could
+  -- just use a setTimeout, but it would not be as prompt. We use a fresh
+  -- counter so that the text change mutation always fires.
+  schedule <- do
+    obsvNode <-
+      DOM.window
+        >>= DOM.document
+        >>> map DOM.htmlDocumentToDocument
+        >>= DOM.createTextNode ""
+        >>> map DOM.textToNode
+    observer <- DOM.mutationObserver \_ _ -> do
+      argsQueue <- readRef argsRef
+      writeRef argsRef []
+      for_ argsQueue notify
+    DOM.observe obsvNode { characterData: true } observer
+    pure \args -> do
+      argsQueue <- readRef argsRef
+      writeRef argsRef (Array.snoc argsQueue args)
+      when (Array.null argsQueue) do
+        fresh ← readRef freshRef
+        writeRef freshRef (fresh + 1)
+        DOM.setNodeValue (show fresh) obsvNode
+
+  let
+    stateFn op state path = do
+      let
+        loc = normalizeLocation state path
+        url = DOM.URL $ unwrap basename <> loc.pathname <> loc.search <> loc.hash
+      DOM.window
+        >>= DOM.history
+        >>= op state (DOM.DocumentTitle "") url
+      schedule loc
+
+    listener =
+      DOM.eventListener \_ -> notify =<< locationState
+
+  DOM.window
+    >>= DOM.windowToEventTarget
+    >>> DOM.addEventListener DOM.popstate listener false
+
+  pure
+    { pushState: stateFn DOM.pushState
+    , replaceState: stateFn DOM.replaceState
+    , locationState
+    , listen
+    }
+
+type LocationState =
+  { state :: Foreign
+  , path :: String
+  , pathname :: String
+  , search :: String
+  , hash :: String
+  }
+
+normalizeLocation :: Foreign -> Path -> LocationState
+normalizeLocation state (Path path) =
```

Comment:
The `state` argument here isn't used. Maybe using `LocationState` w/o `state field here?

Reply:
This is an internal function, so I don't see why I would create another object and type just to extend it again with a new field.

Reply:
There is no comment about this function purpose, so, looking at it I suggested that it takes `state` and do something with it. Even transforming to something like 
```purescript
R.insert (SProxy :: SProxy "state") state case searchIx, hashIx of 
  -- the same as before but w/o state
```
makes it much easier to understand.

Reply:
I don't really see a need to involve `Data.Record` operations since I'm constructing a known shape. However, I have changed the name and formatting though to make it clearer.